osgp/platform/osgp-adapter-ws-smartmetering/src/main/java/org/opensmartgridplatform/adapter/ws/smartmetering/endpoints/SmartMeteringMonitoringEndpoint.java

```diff
@@ -512,12 +516,70 @@ public ClearAlarmRegisterResponse getClearAlarmRegisterResponse(
+    public ActualPowerQualityAsyncResponse getActualPowerQuality(
+            @OrganisationIdentification final String organisationIdentification,
+            @RequestPayload final org.opensmartgridplatform.adapter.ws.schema.smartmetering.monitoring.ActualPowerQualityRequest request, @MessagePriority final String messagePriority,
+            @ResponseUrl final String responseUrl, @ScheduleTime final String scheduleTime) throws OsgpException {
+
+        log.debug("Incoming ActualPowerQualityRequest for meter: {}.", request.getDeviceIdentification());
+
+
+        ActualPowerQualityAsyncResponse response = null;
+
+        try {
+            final ActualPowerQualityRequest dataRequest = this.monitoringMapper
+                    .map(request,
+                            ActualPowerQualityRequest.class);
+
+            final String correlationUid = this.monitoringService.enqueueActualPowerQualityRequestData(
+                    organisationIdentification, request.getDeviceIdentification(), dataRequest, MessagePriorityEnum.getMessagePriority(messagePriority),
+                    this.monitoringMapper.map(scheduleTime, Long.class));
+
+            response = new ActualPowerQualityAsyncResponse();
+            response.setCorrelationUid(correlationUid);
+            response.setDeviceIdentification(request.getDeviceIdentification());
+            this.responseUrlService.saveResponseUrlIfNeeded(correlationUid, responseUrl);
+        } catch (final Exception e) {
+            log.error("Exception: {} while requesting meter reads for device: {} for organisation {}.",
+                    e.getMessage(), request.getDeviceIdentification(), organisationIdentification, e);
+
+            this.handleException(e);
+        }
+        return response;
+    }
+
+    @PayloadRoot(localPart = "ActualPowerQualityAsyncRequest", namespace = SMARTMETER_MONITORING_NAMESPACE)
+    @ResponsePayload
+    public ActualPowerQualityResponse getActualPowerQualityResponse(
+            @OrganisationIdentification final String organisationIdentification,
+            @RequestPayload final ActualPowerQualityAsyncRequest request) throws OsgpException {
+
+        log.debug("Incoming ActualPowerQualityAsyncRequest for meter: {}.", request.getDeviceIdentification());
+
+        ActualPowerQualityResponse response = null;
+        try {
+            final ResponseData responseData = this.responseDataService.dequeue(request.getCorrelationUid(),
+                    ComponentType.WS_SMART_METERING);
+
+            this.throwExceptionIfResultNotOk(responseData, "retrieving actual power data");
+
+            response = this.monitoringMapper.map(responseData.getMessageData(),
+                    org.opensmartgridplatform.adapter.ws.schema.smartmetering.monitoring.ActualPowerQualityResponse.class);
+        } catch (final Exception e) {
+            this.handleRetrieveException(e, request, organisationIdentification);
+        }
+        return response;
+    }
+
 
```

Comment:
Use at most a single blank line for visual separation.

Reply:
changed